etcd-client/src/kv/mod.rs

```diff
@@ -195,26 +260,32 @@ impl Kv {
     /// Will return `Err` if RPC call is failed.
     #[inline]
     pub async fn put(&self, req: EtcdPutRequest) -> Res<EtcdPutResponse> {
-        if self.cache_enable && self.cache_err.load(Ordering::Acquire) {
-            return Err(EtcdError::InternalError("Kv cache meets an error".into()));
-        }
-
+        let inner = self.inner.write().await;
         let key = req.get_key();
         let resp: EtcdPutResponse = retryable!(|| async {
-            let resp = self.client.put_async(&req.clone().into())?;
+            let resp = inner.client.put_async(&req.clone().into())?;
             Ok(From::from(resp.await?))
         });
 
-        if self.cache_enable && self.cache.search(key.clone()).await == None {
-            let revision = resp.get_revision();
-            // Creates a new watch request and adds to the send queue.
-            let mut watch_request = EtcdWatchRequest::create(KeyRange::key(key));
-            watch_request.set_start_revision(revision.cast());
-            if let Err(e) = self.watch_sender.send(watch_request).await {
-                warn!("Fail to send watch request, the error is {}", e);
-                self.cache_err.store(true, Ordering::Release);
+        if inner.cache_enable {
+            if let Some(cache_lock) = inner.cache_availability.try_read() {
+                if inner.cache.search(key.clone()).await == None {
+                    let revision = resp.get_revision();
+                    // Creates a new watch request and adds to the send queue.
+                    let mut watch_request = EtcdWatchRequest::create(KeyRange::key(key));
+                    watch_request.set_start_revision(revision.cast());
+                    if let Err(e) = inner.watch_sender.send(watch_request).await {
+                        warn!(
+                            "Fail to send watch request, the error is {}, clean cache",
+                            e
+                        );
+                        drop(cache_lock);
```

Comment:
Do we need to drop the lock manually here?

Reply:
Because clean_cache will get write lock, so drop read lock here.